### PR TITLE
WIP

### PR DESCRIPTION
**Pull request by jadechoghari**
## What this does
Explain what this PR does. Feel free to tag your PR with the appropriate label(s).

Examples:
|  Title               | Label           |
|----------------------|-----------------|
| Fixes #[issue]       | (🐛 Bug)        |
| Adds new dataset     | (🗃️ Dataset)    |
| Optimizes something  | (⚡️ Performance) |

## How it was tested
Explain/show how you tested your changes.

Examples:
- Added `test_something` in `tests/test_stuff.py`.
- Added `new_feature` and checked that training converges with policy X on dataset/environment Y.
- Optimized `some_function`, it now runs X times faster than previously.

## How to checkout & try? (for the reviewer)
Provide a simple way for the reviewer to try out your changes.

Examples:
```bash
pytest -sx tests/test_stuff.py::test_something
```
```bash
python -m lerobot.scripts.train --some.option=true
```

## SECTION TO REMOVE BEFORE SUBMITTING YOUR PR
**Note**: Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR. Try to avoid tagging more than 3 people.

**Note**: Before submitting this PR, please read the [contributor guideline](https://github.com/huggingface/lerobot/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr).

### CODE REVIEW

File: lerobot/common/datasets/lerobot_dataset.py [hunk 1]
```diff
@@ -177,7 +209,15 @@ def image_keys(self) -> list[str]:
     @property
     def video_keys(self) -> list[str]:
         """Keys to access visual modalities stored as videos."""
-        return [key for key, ft in self.features.items() if ft["dtype"] == "video"]
+        # changed
+        keys = []
+        for key, ft in self.features.items():
+            key_ = (
+                self.inverse_feature_keys_mapping.get(key, key) if self.inverse_feature_keys_mapping else key
```
Comment (danaaubakirova) [inline lerobot/common/datasets/lerobot_dataset.py:217]: why are we doing this?

File: lerobot/common/datasets/lerobot_dataset.py [hunk 2]
```diff
@@ -83,6 +106,7 @@ def __init__(
         root: str | Path | None = None,
         revision: str | None = None,
         force_cache_sync: bool = False,
+        feature_keys_mapping: dict[str, str] | None = None,
```
Comment (danaaubakirova) [inline lerobot/common/datasets/lerobot_dataset.py:110]: Nice! However, I am thinking of the way whether we can handle this outside `lerobot_dataset.py`. Seems it is the best place so far, let's keep it for now:)

File: lerobot/common/datasets/lerobot_dataset.py [hunk 3]
```diff
@@ -717,15 +842,25 @@ def __getitem__(self, idx) -> dict:
...
+        except:
+            print(self.meta.tasks, task_idx, self.repo_id)
+        if "robot_type" not in item:
+            item["robot_type"] = self.robot_type
+        item = map_dict_keys(item, feature_keys_mapping=self.feature_keys_mapping, training_features=self.training_features)
+        # Add padded features
+        # item = self._add_padded_features(item, self.training_features)
+        if self.image_transforms is not None:
+            for cam in item:
+                if cam in self.meta.camera_keys or ("image" in cam and "is_pad" not in cam):
```
Comment (danaaubakirova) [inline lerobot/common/datasets/lerobot_dataset.py:859]: why did you add this?

File: lerobot/common/datasets/lerobot_dataset.py [hunk 4]
```diff
@@ -1022,53 +1256,114 @@ def __init__(
...
-        self.delta_timestamps = delta_timestamps
-        # TODO(rcadene, aliberts): We should not perform this aggregation for datasets
-        # with multiple robots of different ranges. Instead we should have one normalization
-        # per robot.
-        self.stats = aggregate_stats([dataset.meta.stats for dataset in self._datasets])
+        self.delta_timestamps = (
+            delta_timestamps.get(repo_id, None) if delta_timestamps else None
+        )  # delta_timestamps # FIXME(mshukor): last repo?
+        # In case datasets with the same robot_type have different features
+        cleaner = MultiLeRobotDatasetCleaner(
```
Comment (danaaubakirova) [inline lerobot/common/datasets/lerobot_dataset.py:1334]: I would make this optional, also make` MultiLerobotDatasetCleaner` related to smolvla or smolvla2 only. Can we even include it in a separate file and import it if we are doing smolvla2 pretrain?

File: lerobot/common/datasets/lerobot_dataset.py [hunk 5]
```diff
@@ -482,17 +551,62 @@ def __init__(
 
         self.episode_data_index = get_episode_data_index(self.meta.episodes, self.episodes)
 
+        # mustafa code
+        if self.discard_first_n_frames > 0:
```
Comment (danaaubakirova) [inline lerobot/common/datasets/lerobot_dataset.py:556]: honestly, we used this only for some experiments, I don't think we should include it to the basic LeRobotDataset Class for now. 

Maybe let's add #FIXME or #TODO to remove it in the future? Since we didn't use this feature in Smolvla anyways.

File: lerobot/common/datasets/lerobot_dataset.py [hunk 6]
```diff
@@ -482,17 +551,62 @@ def __init__(
...
+        # Mustafa
+        self.meta.info["features"] = map_dict_keys(
+            self.meta.info["features"], feature_keys_mapping=self.feature_keys_mapping, training_features=self.training_features
+        )
+        self.keys_to_max_dim = {
+            ACTION: max_action_dim,
+            OBS_ENV_STATE: max_state_dim,
+            OBS_STATE: max_state_dim,
+            OBS_IMAGE: max_image_dim,
+            OBS_IMAGE_2: max_image_dim,
```
Comment (danaaubakirova) [inline lerobot/common/datasets/lerobot_dataset.py:599]: Sometimes we pass only 2 observation images, or 1. Will this still work?

File: lerobot/common/datasets/lerobot_dataset.py (hunk 2 above)
Comment (AdilZouitine) [reply lerobot/common/datasets/lerobot_dataset.py:110]: Maybe my PR #1431 will solve this, maybe keep it as it is, and when merging my PR use processor instead

Comment (jadechoghari) [reply lerobot/common/datasets/lerobot_dataset.py:110]: At some point yes we should move it away

File: lerobot/common/datasets/lerobot_dataset.py (hunk 3 above)
Comment (jadechoghari) [reply lerobot/common/datasets/lerobot_dataset.py:859]: I'll investigate if we can go without it, i think he's avoiding transforming 'fake' padded images

File: lerobot/common/datasets/lerobot_dataset.py (hunk 4 above)
Comment (jadechoghari) [reply lerobot/common/datasets/lerobot_dataset.py:1334]: Yes maybe also updating the naming to reflect its relation to smolvla

File: lerobot/common/datasets/lerobot_dataset.py (hunk 6 above)
Comment (jadechoghari) [reply lerobot/common/datasets/lerobot_dataset.py:599]: It should, I'll add a test for this!
